iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/DirectoryCopySourceCheck.java

```diff
@@ -0,0 +1,121 @@
+      String path = SyntaxTokenUtils.trimmedQuotesValues(src);
+      if (!path.startsWith("http://") && !path.startsWith("https://")) {
+        reportIfSensitive(ctx, src, isSensitivePath(path), "ADD");
+      }
+    }
+  }
+
+  private static void checkCopy(CheckContext ctx, CopyTree copy) {
+    if (hasOption(copy.options(), "from")) return;
+
+    for (SyntaxToken src : copy.srcs()) {
+      reportIfSensitive(ctx, src, isSensitivePath(SyntaxTokenUtils.trimmedQuotesValues(src)), "COPY");
+    }
+  }
+
+  private static boolean hasOption(List<ParamTree> options, String key) {
+    return options.stream().anyMatch(param -> param.name().equals(key));
+  }
+
+  private static void reportIfSensitive(CheckContext ctx, SyntaxToken src, PathSensitivity sensitivity, String instructionName) {
+    if(sensitivity == PathSensitivity.ROOT_OR_CURRENT) {
+      ctx.reportIssue(src, MESSAGE_CURRENT_OR_ROOT);
+    } else if (sensitivity == PathSensitivity.TOP_LEVEL_GLOBBING) {
+      ctx.reportIssue(src, String.format(MESSAGE_GLOBING, instructionName));
+    }
+  }
+
+  private enum PathSensitivity {
+    SAFE, ROOT_OR_CURRENT, TOP_LEVEL_GLOBBING
+  }
+
+  /**
+   * Indicate if a path is sensitive. A path is sensitive if any of the current condition is true :
+   * <ul>
+   * <li>if the provided path resolve to a root level or to the current directory</li>
+   * <li>if there is any top level entry (level <= 1) ending with a wildcard</li>
+   * </ul>
+   * Examples of root level : '/' ; '/test/..' ; 'c:/' ; 'c:/test/..'
+   * Examples of current level : '.' ; './test/..'
+   * Examples of top level entry ending with wildcard : 'a*' ; './a*' ; '/a*' ; './test/../a*'
+   */
+  private static PathSensitivity isSensitivePath(String path) {
+    String[] splitted = path.split("/");
+    String regexWindowsDrive = "^[a-zA-Z]:$";
+    boolean isAbsolutePath = splitted.length == 0 || splitted[0].isEmpty() || splitted[0].matches(regexWindowsDrive);
+    int level = 0;
+    boolean hasTopLevelWithWildcard = false;
+
+    for (int i = 0; i < splitted.length; i++) {
+      if (splitted[i].equals("..")) {
+        level--;
+      } else if (!splitted[i].equals(".") && !splitted[i].isEmpty() && (i > 0 || !splitted[i].matches(regexWindowsDrive))) {
+        level++;
+      }
+
+      if (level <= 1 && splitted[i].endsWith("*")) {
+        hasTopLevelWithWildcard = true;
+      }
+
+      if (level < 0) {
+        if (isAbsolutePath) {
+          // in case of absolute path, we can't go above root, so we reset to root
+          level = 0;
+        } else {
+          // in case of relative path, if we go above the current folder, we suppose we lose the track and can't report any issue
+          return PathSensitivity.SAFE;
+        }
+      }
+    }
+
+    if (level == 0) return PathSensitivity.ROOT_OR_CURRENT;
+    if (level == 1 && hasTopLevelWithWildcard) return PathSensitivity.TOP_LEVEL_GLOBBING;
+    return PathSensitivity.SAFE;
```

Comment:
As discussed:
```suggestion
  private static PathSensitivity isSensitivePath(String path) {
    String[] levels = normalize(path);
    if (levels.length == 0) return PathSensitivity.ROOT_OR_CURRENT;
    if (levels.length == 1 && (isRootOrCurrent(levels[0]))) return PathSensitivity.ROOT_OR_CURRENT;
    if (Arrays.stream(levels).limit(2).anyMatch(l -> l.endsWith("*"))) return PathSensitivity.TOP_LEVEL_GLOBBING;
    return PathSensitivity.SAFE;
  }

  private static boolean isRootOrCurrent(String level) {
    return (level.isEmpty() || ".".equals(level) || WINDOWS_DRIVE_PATTERN.matcher(level).find());
  }


  private static String[] normalize(String path) {
    Deque<String> levels = new ArrayDeque<>();
    for (String current : path.split("/")) {
      if ("..".equals(current) && !levels.isEmpty()) {
        levels.removeLast();
      } else if (current.isEmpty()) {
        if (levels.isEmpty()) {
          levels.add(current);
        }
      } else if (".".equals(current) && levels.isEmpty()) {
        levels.add(current);
      } else if (!".".equals(current)) {
        levels.add(current);
      }
    }
    return levels.toArray(new String[] {});
  }
```

Reply:
Wonderful piece of code ! And as you saw, it even remove FP/FN of the previous code ^^'
Thanks !

Reply:
woohoo!